target/linux/generic/hack-5.10/780-1-net-dsa-allow_for_multiple_CPU_ports.patch

```diff
@@ -0,0 +1,216 @@
+Cc:     Andrew Lunn <andrew@lunn.ch>,
+        Vivien Didelot <vivien.didelot@gmail.com>,
+        Florian Fainelli <f.fainelli@gmail.com>,
+        David Ahern <dsahern@gmail.com>,
+        Stephen Hemminger <stephen@networkplumber.org>,
+        =?UTF-8?q?Marek=20Beh=C3=BAn?= <marek.behun@nic.cz>
+Subject: [PATCH RFC net-next 1/3] net: dsa: allow for multiple CPU ports
+Date:   Sat, 24 Aug 2019 04:42:48 +0200
+Message-Id: <20190824024251.4542-2-marek.behun@nic.cz>
+X-Mailer: git-send-email 2.21.0
+In-Reply-To: <20190824024251.4542-1-marek.behun@nic.cz>
+References: <20190824024251.4542-1-marek.behun@nic.cz>
+MIME-Version: 1.0
+Content-Type: text/plain; charset=UTF-8
+Content-Transfer-Encoding: 8bit
+X-Virus-Scanned: clamav-milter 0.100.3 at mail.nic.cz
+X-Virus-Status: Clean
+Sender: netdev-owner@vger.kernel.org
+Precedence: bulk
+List-ID: <netdev.vger.kernel.org>
+X-Mailing-List: netdev@vger.kernel.org
+Archived-At: <https://lore.kernel.org/netdev/20190824024251.4542-2-marek.behun@nic.cz/>
+List-Archive: <https://lore.kernel.org/netdev/>
+List-Post: <mailto:netdev@vger.kernel.org>
+
+Allow for multiple CPU ports in a DSA switch tree. By default assign the
+CPU ports to user ports in a round robin way, ie. if there are two CPU
+ports connected to eth0 and eth1, and five user ports (lan1..lan5),
+assign them as:
+  lan1 <-> eth0
+  lan2 <-> eth1
+  lan3 <-> eth0
+  lan4 <-> eth1
+  lan5 <-> eth0
+
+Signed-off-by: Marek Behún <marek.behun@nic.cz>
+---
+ include/net/dsa.h |  5 +--
+ net/dsa/dsa2.c    | 84 +++++++++++++++++++++++++++++++----------------
+ 2 files changed, 58 insertions(+), 31 deletions(-)
+
+diff --git a/net/dsa/dsa2.c b/net/dsa/dsa2.c
+index 8c4eccb0cfe6..c5af89079a6b 100644
+@@ -211,36 +212,43 @@ static bool dsa_tree_setup_routing_table(struct dsa_switch_tree *dst)
+ 	return complete;
+ }
+ 
+-static struct dsa_port *dsa_tree_find_first_cpu(struct dsa_switch_tree *dst)
++static int dsa_tree_setup_default_cpus(struct dsa_switch_tree *dst)
+ {
+-	struct dsa_port *dp;
+-
++	struct dsa_port *cpu_dp, *dp, *first_cpu_dp = NULL, *last_cpu_dp = NULL;
++	
++	/* Find first and last CPU port */
+ 	list_for_each_entry(dp, &dst->ports, list)
+-		if (dsa_port_is_cpu(dp))
+-			return dp;
+-
+-	return NULL;
+-}
++		if (dsa_port_is_cpu(dp)) {
++			if (first_cpu_dp == NULL)
++				first_cpu_dp = dp;
+ 
+-static int dsa_tree_setup_default_cpu(struct dsa_switch_tree *dst)
+-{
+-	struct dsa_port *cpu_dp, *dp;
++			last_cpu_dp = dp;
++		}
+ 
+-	cpu_dp = dsa_tree_find_first_cpu(dst);
+-	if (!cpu_dp) {
++	if (first_cpu_dp == NULL) {
+ 		pr_err("DSA: tree %d has no CPU port\n", dst->index);
+ 		return -EINVAL;
+ 	}
+ 
+-	/* Assign the default CPU port to all ports of the fabric */
++	cpu_dp = first_cpu_dp;
++
++	/* Assign the CPU ports in round-robbin way to all ports of the fabric */
+ 	list_for_each_entry(dp, &dst->ports, list)
+-		if (dsa_port_is_user(dp) || dsa_port_is_dsa(dp))
++		if (dsa_port_is_user(dp) || dsa_port_is_dsa(dp)) {
```

Comment:
As proposed by @frank-w, we can use dts to set the default CPU port, and fallback to round-robin if not present.

```diff
 	list_for_each_entry(dp, &dst->ports, list)
-		if (dsa_port_is_user(dp) || dsa_port_is_dsa(dp))
+		if (dsa_port_is_user(dp) || dsa_port_is_dsa(dp)) {
+			dp->cpu_dp = dsa_tree_find_port_by_node(dst,
+					of_parse_phandle(dp->dn, "default_cpu", 0));
+			if (dp->cpu_dp && dsa_port_is_cpu(dp->cpu_dp))
+				continue;
+
 			dp->cpu_dp = cpu_dp;
```

Then we can define the `default_cpu` property in dts, for example:
```dts
&switch0 {
	ports {
		port@0 {
			status = "okay";
			label = "lan1";
			default_cpu = <&dsa_cpu0>;
		};

		port@1 {
			status = "okay";
			label = "lan2";
			default_cpu = <&dsa_cpu0>;
		};

		port@2 {
			status = "okay";
			label = "lan3";
			default_cpu = <&dsa_cpu0>;
		};

		port@4 {
			status = "okay";
			label = "wan";
			default_cpu = <&dsa_cpu1>;
		};

		dsa_cpu1: port@5 {
			reg = <5>;
		};

		dsa_cpu0: port@6 {
			reg = <6>;
		};
```